middleman_protocol/middleman_protocol/message.py

```diff
@@ -0,0 +1,387 @@
+from .exceptions import PayloadTypeInvalidMiddlemanProtocolError
+from .exceptions import RequestIdInvalidTypeMiddlemanProtocolError
+from .exceptions import SignatureInvalidMiddlemanProtocolError
+from .registry import PAYLOAD_TYPE_TO_MIDDLEMAN_MESSAGE_CLASS
+from .registry import register
+
+
+class AbstractMiddlemanMessage(ABC):
+    """
+    Abstract class for inheriting by Middleman protocol messages.
+
+    Contains base functionality like:
+    - serialization,
+    - deserialization,
+    - frame validation.
+
+    Subclasses must implement:
+    - parsing payload to bytes on serialization,
+    - parsing serialized bytes payload back to original format on deserialization,
+    - validation of original payload before serialization.
+
+    """
+
+    __slots__ = [
+        'payload',
+        'payload_type',
+        'request_id',
+    ]
+
+    def __init__(self, payload: Any, request_id: int) -> None:
+        self._validate_request_id(request_id)
+        self.request_id = request_id
+
+        self._validate_payload(payload)
+        self.payload = payload
+
+    @classmethod
+    @abstractmethod
+    def _deserialize_payload(cls, payload: bytes):
+        pass
+
+    @abstractmethod
+    def _serialize_payload(self, payload: Any):
+        pass
+
+    @abstractmethod
+    def _validate_payload(self, payload: Any):
+        pass
+
+    @classmethod
+    def get_frame_format(cls) -> Struct:
+        """ Returns Struct object containing frame structure. """
+        frame_format = Struct(
+            frame_signature=Bytes(FRAME_SIGNATURE_BYTES_LENGTH),
+            signed_part_of_the_frame=Struct(
+                request_id=BytesInteger(FRAME_REQUEST_ID_BYTES_LENGTH),
+                payload_type=Enum(Byte, PayloadType),
+                payload_length=BytesInteger(FRAME_PAYLOAD_LENGTH_BYTES_LENGTH),
+                payload=GreedyBytes,
+            ),
+        )
+
+        assert FRAME_PAYLOAD_LENGTH_STARTING_BYTE == (
+            frame_format.frame_signature.subcon.length +
+            frame_format.signed_part_of_the_frame.request_id.subcon.length +
+            1  # 1 byte for payload_type
+        )
+        assert FRAME_PAYLOAD_STARTING_BYTE == (
+            frame_format.frame_signature.subcon.length +
+            frame_format.signed_part_of_the_frame.request_id.subcon.length +
+            1 +  # 1 byte for payload_type
+            frame_format.signed_part_of_the_frame.payload_length.subcon.length
+        )
+
+        return frame_format
+
+    @classmethod
+    def add_frame_separator(cls, raw_frame: bytes) -> bytes:
+        """ Adds frame separator to raw frame. """
+        assert isinstance(raw_frame, bytes)
+
+        return raw_frame + FRAME_SEPARATOR
+
+    @classmethod
+    def remove_frame_separator(cls, raw_message: bytes) -> bytes:
+        """ Remove frame separator from raw frame. """
+        assert isinstance(raw_message, bytes)
+        assert FRAME_SEPARATOR in raw_message
```

Comment:
This won't work correctly. Consider this example:

|                                      |               |
|--------------------------------------|---------------|
| Separator                            | `66 66 66`
| Escape character                     | `33 33 33`
| Escaped separator                    | `33 33 33 00`
| Escaped escape character             | `33 33 33 01`
|                                      |               |
| Input                                | `22 33 33 33 00 66 66 66`
| Output from `handle_frame_receive()` | `22 66 66 66`

So if someone uses your function, it will receive `22 66 66 66`. But this `66 66 66` is not a separator. It's a part of the frame itself.

This function would be useful only if someone got escaped input but without separators removed. I can't really think of any situation where it could happen in practice because the removal happens before unescaping.

Reply:
Yes, with new implementation it can be removed. 

But check your example again, it actually gives proper output.

Reply:
I'm not saying that it gives you wrong output as used currently. I'm saying that it's dangerous because someone could assume that it can detect the separator and that it would be safe to use it on a frame twice. If he did that, he would get `22` which is not the full frame.